### ISSUE TITLE
実装: RuleBased/Conservative 50:50環境で最終役職と自陣営獲得枚数を教師に競り価値を学習する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#375]
Related: #202, #367, #368, #370, #372, #373
Base branch: `feature/issue-202-non-playing-ai`

## 背景

Issue #373 で最終役職へ分解した競りQ学習を試したが、現行の Conservative / All-Pass 50:50 環境では、強制した BID sample `5579/5579` がすべて candidate Napoleon になり、役職ラベルは PASS-vs-BID の粗い区別しか提供しなかった。

また、Issue #372 では terminal reward が主に `-1 / 0 / 大きな正値` に集中し、同一stateのsuit/target間で exact tie が多いことが分かった。

次は、対戦環境をより通常の競りに近づけたうえで、単純な成功/失敗rewardではなく、**最終役職と、その役職で所属する陣営が最終的に何枚のポイント札/絵札を獲得したか** を教師として学習する。

## 方針

Frozen bidding opponents を初版の通常構成へ戻す。

- RuleBased 50%
- Conservative 50%
- frozen seatごとにseeded deterministicで独立選択

All-Passは今回の主要学習相手から外す。

理由:

- 強制BIDすると必ずcandidate Napoleonになる極端な分布を避ける
- 他playerのraiseによりcandidateが競り負け、市民/副官になる展開も教師へ含める
- actionが最終役職や競り結果へ与える影響を通常の対戦分布で学習する

## 目的

競りstate `s` とaction `a` に対して、少なくとも以下を予測できるモデルを作る。

1. `P(finalRole | s,a)`
2. `E(teamPointCards | s,a, finalRole)`

最終役職 `finalRole` は教師ラベルであり、推論時には未知なので入力特徴へ入れてはいけない。

概念的には、actionごとに

```text
role distribution
+
role-conditioned team point-card prediction
```

を持つ。

このIssueではまず、これらの予測がstate/actionに応じた細かいsignalを持つかを検証する。

## 自陣営獲得枚数の定義

個人の獲得枚数ではなく、**candidateが最終的に所属した陣営全体の獲得ポイント札枚数**を教師にする。

例:

- candidate finalRole = Napoleon
  - teacher = Napoleon側が最終的に獲得したポイント札枚数
- candidate finalRole = Adjutant
  - teacher = Napoleon側が最終的に獲得したポイント札枚数
- candidate finalRole = Citizen
  - teacher = 連合軍側が最終的に獲得したポイント札枚数
- NoContract / AllPass
  - 通常の獲得枚数教師は存在しないので、別mask / nullable / dedicated classとして扱う

既存game-coreで「ポイント札/絵札」がどのカウントに相当するかを確認し、ルール上の宣言枚数13..19と直接比較可能な枚数を使うこと。

## Dataset schema拡張

Issue #368 counterfactual Q dataset schemaを必要最小限拡張し、raw sampleごとに最低限以下を保存する。

- finalRole
- candidate final alliance/team
- Napoleon-side point-card count
- Coalition-side point-card count
- candidate-team point-card count
- declared target（最終Napoleon contractが存在する場合）
- contract success/failure
- contract margin（意味がある場合）

既存schemaとの互換性/ versioningを明確にする。

NoContract時に架空の0枚を教師として入れて通常回帰を歪めないこと。

## 対戦環境

counterfactual source / continuationでは frozen bidding opponents を以下にする。

- RuleBased / Conservative = 50:50
- per-seat independently sampled
- seeded deterministic

Adjutant / exchange / playing は現行frozen topologyを維持する。

playingは既存 `ppo-separated-v1000` を使用する。

opponent mix / artifacts / SHAをmanifestへ保存する。

## Counterfactual action coverage

Issue #368の方式を継続する。

各stateで少なくとも:

- PASS
- lowest legal S/H/D/C
- strongest-suit candidate
- policy top candidate
- random/target coverage candidates

を含める。

全29 actionを毎state必須にはしないが、S/H/D/Cおよびtarget 13..19 coverageをsummaryで確認する。

## Role probability model

Issue #373のrole-decomposed構造を再利用・改善してよい。

例:

```text
compact278 -> shared MLP -> 29 x R role logits
```

教師:

`actual finalRole`

validationでは、majority baseline / action-index role-frequency baselineと比較し、state情報が効いているか確認する。

特に今回、RuleBased相手のraiseによりBID後にもCitizen/Adjutant等が発生するかを最初にdataset診断する。

## Team point-card prediction

役職条件付きで、candidate teamの最終ポイント札枚数を予測する。

候補構造:

```text
compact278 -> shared/value MLP -> 29 x R teamPointCards
```

各sampleでは、forced action と actual finalRole に対応する1セルのみloss対象にする。

NoContractは通常回帰loss対象外にする。

lossはまず MSE / Huber を比較可能にし、枚数は自然な整数範囲のraw targetを基本とする。

## 追加target / diagnostics

teamPointCardsから以下を派生させ、教師signalを確認する。

Napoleon/Adjutantの場合:

```text
contractMargin = napoleonSidePointCards - declaredTarget
```

Citizenの場合は宣言targetとの差を同じ意味で扱わないこと。

最低限diagnosticsとして:

- role別 teamPointCards distribution
- action別 teamPointCards mean/std
- same-state suit間のteamPointCards差
- same-state target間のteamPointCards差
- strongest suit bid vs other suits
- finalRole別 prediction MAE/RMSE

を出す。

### 仮説A: opponent mix改善

RuleBased / Conservative 50:50にすることで、BID後のfinalRoleがNapoleon一色ではなくなるか。

必ず以下を報告:

- PASS後role distribution
- BID後role distribution
- suit別role distribution
- target別role distribution

### 仮説B: 枚数教師signal

同じstateで13S/13H/13D/13C等を試したとき、最終teamPointCardsに差が出るか。

最低限:

- same-target 4-suit complete statesのbest-worst gap
- exact tie率
- strongest suitがteamPointCards bestになる率

をIssue #372のterminal reward教師と比較する。

### 仮説C: target signal

同じsuitでtargetを変えたとき、teamPointCards / contract marginに学習可能な差があるか。

### Role

- accuracy
- macro F1
- per-role precision/recall
- confusion matrix
- majority baseline
- action-index baseline

### Team point cards

- MAE
- RMSE
- correlation
- role別MAE/RMSE
- suit別MAE/RMSE
- target別MAE/RMSE

### Action ranking

teacher-covered actions内で、用途別に評価する。

Napoleon側になったaction比較では teamPointCards / contract marginを用いる。

異なるfinalRole間のactionをどう最終価値へ統合するかは、このIssueで無理に固定しなくてもよい。

まず:

- role predictionが成立するか
- role-conditioned teamPointCards predictionが成立するか
- suit/target間の連続教師signalが増えたか

を確認する。

## Small experiment

実装後、推奨:

- 1000 source states程度
- candidate 5-seat rotation
- RuleBased / Conservative 50:50 frozen bidding opponents
- repeats=1（必要なら2）
- PASS + lowest legal S/H/D/C + target補完

でdatasetを生成し、small training experimentを実行する。

## Success criteria

以下が確認できればこの方向を継続する。

- BID後roleがNapoleon一色ではなくなる
- action-index baselineを超えてstate-conditioned role予測ができる
- teamPointCards教師がterminal reward教師よりsuit/targetのexact tieを大幅に減らす
- teamPointCards predictionに明確な正相関がある
- strongest suitとbest suitの関係がランダムより改善
- target変更による意味のある連続signalが確認できる

## 次段階

このIssueでは最終的なruntime action scoreの設計を急がない。

結果が良ければ後続Issueで、例えば

```text
P(role|s,a)
+
E(teamPointCards|s,a,role)
+
contract target
```

から実際の競りaction選択scoreを設計する。

もしteamPointCards予測でもsignalが弱い場合は、playing policy / state selection / observation自体を再検討する。

## Tests

最低限:

1. RuleBased / Conservative 50:50 seeded selection
2. dataset schema versioning
3. finalRole extraction
4. candidate alliance/team extraction
5. Napoleon-side point-card count fixture
6. Coalition-side point-card count fixture
7. candidate-team point-card count fixture
8. NoContractはpoint-card regressionからmask
9. finalRole future leakageなし
10. role head output/loss
11. teamPointCards head output/loss
12. stateKey split leakageなし
13. deterministic CPU smoke
14. summary/diagnostics correctness
15. checkpoint save/load
16. ONNX exportする場合はPyTorch parity

## Git / PR運用

最新 `feature/issue-202-non-playing-ai` からIssue専用ブランチを作成する。

PR baseは必ず `feature/issue-202-non-playing-ai`。
main向けPRは作らない。

実装・tests・small experiment後、Codex review最大4回（review後修正最大3回）。blocking/actionable指摘がなくなればReadyにして親ブランチへmergeする。

mainにはmergeしない。

## 完了時に必ず報告

- dataset schema/version
- opponent mix
- dataset state/sample数
- PASS/BID別finalRole distribution
- suit/target別finalRole distribution
- teamPointCards distribution
- same-state suit gap / exact tie率
- strongest suit teacher-best率
- target差
- role classifier metrics + baselines
- teamPointCards prediction MAE/RMSE/correlation
- role別/suit別/target別metrics
- この教師が#372/#373より明確に改善したか
- 次にaction selection score設計へ進むべきか
- PR / merge commit / Codex review回数
- mainへmergeしていないこと
